processor/lsmintervalprocessor/internal/merger/limits/tracker.go

```diff
@@ -0,0 +1,297 @@
+func (t *Tracker) Marshal() ([]byte, error) {
+	// TODO (lahsivjar): Estimate the size of the trackers to optimize
+	// allocations. Also see: https://github.com/axiomhq/hyperloglog/issues/44
+	b := make([]byte, 8) // reserved for observed count
+
+	var offset int
+	binary.BigEndian.PutUint64(b[offset:(offset+8)], t.observedCount)
+	offset += 8
+	if t.overflowCounts != nil {
+		hll, err := t.overflowCounts.MarshalBinary()
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal limits: %w", err)
+		}
+		b = slices.Grow(b, len(hll))[:offset+len(hll)]
+		copy(b[offset:], hll)
+	}
+	return b, nil
+}
+
+// Unmarshal unmarshals the encoded limits to go struct. Example usage:
+//
+//	var t Tracker
+//	if err := t.Unmarshal(data); err != nil {
+//	    panic(err)
+//	}
+func (t *Tracker) Unmarshal(d []byte) error {
+	if len(d) < 8 {
+		return errors.New("failed to unmarshal tracker, invalid length")
+	}
+
+	var offset int
+	t.observedCount = binary.BigEndian.Uint64(d[offset:(offset + 8)])
+	offset += 8
+	if len(d) == offset {
+		return nil
+	}
+	t.overflowCounts = hyperloglog.New14()
+	if err := t.overflowCounts.UnmarshalBinary(d[offset:]); err != nil {
+		return fmt.Errorf("failed to unmarshal tracker: %w", err)
+	}
+	return nil
+}
+
+type trackerType uint8
+
+const (
+	resourceTracker trackerType = iota
+	scopeTracker
+	dpsTracker
+)
+
+// Trackers represent multiple tracker in an ordered structure. It takes advantage
+// of the fact that pmetric DS is ordered and thus allows trackers to be created
+// for each resource, scope, and datapoint independent of the pmetric datastructure.
+// Note that this means that the order for pmetric and trackers are implicitly
+// related and removing/adding new objects to pmetric should be accompained by
+// adding a corresponding tracker.
+type Trackers struct {
+	resourceLimit uint64
+	scopeLimit    uint64
+	scopeDPLimit  uint64
+
+	resource *Tracker
+	scope    []*Tracker
+	scopeDPs []*Tracker
+}
+
+func NewTrackers(resourceLimit, scopeLimit, scopeDPLimit uint64) *Trackers {
+	return &Trackers{
+		resourceLimit: resourceLimit,
+		scopeLimit:    scopeLimit,
+		scopeDPLimit:  scopeDPLimit,
+
+		// Create a resource tracker preemptively whenever a tracker is created
+		resource: newTracker(resourceLimit),
+	}
+}
+
+func (t *Trackers) NewScopeTracker() *Tracker {
+	newTracker := newTracker(t.scopeLimit)
+	t.scope = append(t.scope, newTracker)
+	return newTracker
+}
+
+func (t *Trackers) NewScopeDPsTracker() *Tracker {
+	newTracker := newTracker(t.scopeDPLimit)
+	t.scopeDPs = append(t.scopeDPs, newTracker)
+	return newTracker
+}
+
+func (t *Trackers) GetResourceTracker() *Tracker {
+	return t.resource
+}
+
+func (t *Trackers) GetScopeTracker(i int) *Tracker {
+	if i >= len(t.scope) {
+		return nil
+	}
+	return t.scope[i]
+}
+
+func (t *Trackers) GetScopeDPsTracker(i int) *Tracker {
+	if i >= len(t.scopeDPs) {
+		return nil
+	}
+	return t.scopeDPs[i]
+}
+
+func (t *Trackers) Marshal() ([]byte, error) {
```

Comment:
Should we add a version byte (first byte) for future proofing?

Reply:
Done in https://github.com/elastic/opentelemetry-collector-components/pull/235/commits/4446f97f592f931916dbc9c2e146e1737a54babe